crates/fuel-streams-core/src/transactions/types.rs

```diff
@@ -1,27 +1,426 @@
 
-#[cfg(any(test, feature = "test-helpers"))]
-impl MockTransaction {
-    pub fn build() -> Transaction {
-        Transaction::default_test_tx()
+impl Transaction {
+    pub fn new(
+        id: &Bytes32,
+        transaction: &FuelCoreTransaction,
+        status: &TransactionStatus,
+        base_asset_id: &FuelCoreAssetId,
+    ) -> Self {
+        let bytecode_root = {
+            use fuel_core_types::fuel_tx::field::BytecodeRoot;
+            match transaction {
+                FuelCoreTransaction::Upload(tx) => {
+                    Some((*tx.bytecode_root()).into())
+                }
+                _ => None,
+            }
+        };
+
+        let bytecode_witness_index = {
+            use fuel_core_types::fuel_tx::field::BytecodeWitnessIndex;
+            match transaction {
+                FuelCoreTransaction::Upload(tx) => {
+                    Some(*tx.bytecode_witness_index())
+                }
+                _ => None,
+            }
+        };
+
+        let blob_id = {
+            use fuel_core_types::fuel_tx::field::ChargeableBody;
+            match transaction {
+                FuelCoreTransaction::Blob(blob) => Some(blob.body().id.into()),
+                _ => None,
+            }
+        };
+
+        let input_asset_ids = {
+            use fuel_core_types::fuel_tx::Executable;
+
+            match transaction {
+                FuelCoreTransaction::Script(tx) => Some(
+                    tx.input_asset_ids(base_asset_id)
+                        .map(|c| AssetId::from(*c))
+                        .collect(),
+                ),
+                FuelCoreTransaction::Create(tx) => Some(
+                    tx.input_asset_ids(base_asset_id)
+                        .map(|c| AssetId::from(*c))
+                        .collect(),
+                ),
+                FuelCoreTransaction::Mint(_) => None,
+                FuelCoreTransaction::Upgrade(tx) => Some(
+                    tx.input_asset_ids(base_asset_id)
+                        .map(|c| AssetId::from(*c))
+                        .collect(),
+                ),
+                FuelCoreTransaction::Upload(tx) => Some(
+                    tx.input_asset_ids(base_asset_id)
+                        .map(|c| AssetId::from(*c))
+                        .collect(),
+                ),
+                FuelCoreTransaction::Blob(tx) => Some(
+                    tx.input_asset_ids(base_asset_id)
+                        .map(|c| AssetId::from(*c))
+                        .collect(),
+                ),
+            }
+        };
+
+        let input_contract = {
+            use fuel_core_types::fuel_tx::field::InputContract;
+            match transaction {
+                FuelCoreTransaction::Mint(mint) => {
+                    Some(mint.input_contract().into())
+                }
+                _ => None,
+            }
+        };
+
+        let input_contracts = {
+            use std::vec::IntoIter;
+
+            fn input_contracts<Tx>(
```

Comment:
This function is being created on each new `Transaction::new()` execution; since it's a pure function, I guess you could add it as a `Transaction` method or get it out of the this scope

Reply:
And if you just do this is not the same?

```rust
        let input_contracts = {
            match transaction {
                FuelCoreTransaction::Mint(_) => None,
                tx => {
                    let mut inputs: Vec<_> = tx
                        .inputs()
                        .iter()
                        .filter_map(|input| match input {
                            fuel_tx::Input::Contract(contract) => {
                                Some(contract.contract_id)
                            }
                            _ => None,
                        })
                        .collect();
                    inputs.sort();
                    inputs.dedup();
                    Some(inputs.into_iter().map(|id| (*id).into()).collect())
                }
            }
        };
```

Reply:
Nice 🚀. Will update.

Reply:
Done ✅